Initial draft of Homebrew migration guide (#96)
flox/floxdocs: docs/cookbook/migrations/homebrew.md

---
title: Homebrew
description: Common questions and solutions for using Rust with Flox
---

# Homebrew migration guide
Flox can replace Homebrew entirely, or they can be used together.

This guide explains how to introduce Flox into environments where Homebrew is currently being used, either as a replacement or an addition. It introduces new concepts and proposes a basic procedure for mapping packages.

## Why you might want to migrate
Homebrew does a great job, and has been loved as the “missing package manager” for a generation of macOS users. But there are a few reasons you might consider moving to Flox:

* **You want to install packages into discrete environments.** Flox allows you to create as many environments as you need, each with a different set of packages.

> * **You need a reproducible manifest.** Flox environments are defined in a TOML manifest that can be shared and reproduced - often across OS and system architectures.

Reviewer: Edit: nvm, I see the `leaves` command below.

> * **You need a reproducible manifest.** Flox environments are defined in a TOML manifest that can be shared and reproduced - often across OS and system architectures.

Author: I didn't go with your exact wording but it did make me realize we aren't highlighting the "take it anywhere" aspect. Let me know what you think of the current text - does it say enough now?

* **You need a broader set of packages.** Homebrew has a sizeable collection of packages, but the Flox Catalog is based on Nixpkgs, the largest collection of packages in the world.
* **You need older versions of packages.** The Flox Catalog contains historical versions of each of its packages - in many cases, going back decades.

## New concepts

> Flox works similarly to Homebrew, but there are a few new concepts to be aware of.

Reviewer: I would say "Flox feels similar to Homebrew" since it doesn't actually work anything like Homebrew under the hood. I think the similarities are mainly "you have install, uninstall, and search commands".

> Flox works similarly to Homebrew, but there are a few new concepts to be aware of.

Reviewer 2: "new concepts to be aware of" feels close to "be wary of" in this context. Could this section be flipped around to explain the happy path of the default env _first_ and then get into how all flox environments are virtual envs that need to be init and activated (but it was automatic and special for the default one)?

Author: OK - I restructured the doc and I feel a lot better about it now. Let me know if you like the new flow.

Reviewer 2: Yes! Love it! That takes care of my concerns, and it flows much better overall imo.

### Virtual environments

As a Homebrew user, you will find several of the Flox subcommands familiar:

* `search` is used to find available packages
* `install` is used to install packages
* `uninstall` is used to remove packages

However, there is a fundamental design difference: **Flox installs packages into virtual environments**. These environments live inside directories, and you can have as many as you want. Packages are placed into a central store and mapped into each of your environments based on their needs.

That means there are a few additional subcommands to learn when using Flox:

* `init` creates a new environment

> * `activate` activates an environment

Reviewer: ```suggestion
* `activate` activates an environment to make the packages available
```

Just in case "activates an environment" doesn't convey enough information.

### The default environment
Typically, a Flox user will create a [default environment]() in their home directory that is activated each time a new shell is created. This is useful for packages that should be available to the user across all of the contexts where they work.

> When Flox is configured with a default environment, it behaves very similarly to Homebrew. The Flox CLI will assume the default environment when you call `flox install` commands.

Reviewer: ```suggestion
When Flox is configured with a default environment, it behaves very similarly to Homebrew. The Flox CLI will assume the default environment when you run `flox install` in a directory that doesn't contain an environment of its own.
```

For example:

```
% brew leaves
boost
gh
imagemagick
mysql
openssh
```

Then, search Flox for one of the formulae you have installed with Homebrew. In this case, for example, you could choose `boost`:

```
% flox search boost
boost     Collection of C++ libraries
booster   Fast and secure initramfs generator
xgboost   Scalable, Portable and Distributed Gradient Boosting (GBDT, GBRT or GBM) Library
boost155  Collection of C++ libraries
boost159  Collection of C++ libraries
boost15x  Collection of C++ libraries
boost160  Collection of C++ libraries
boost165  Collection of C++ libraries
boost166  Collection of C++ libraries
boost167  Collection of C++ libraries

Showing 10 of 120 results. Use `flox search boost --all` to see the full list.

Use 'flox show <package>' to see available versions
```

The first one on the list is the correct Flox package to install, and it has the same name as the Homebrew package. This will often be the case, but not always.

### Install your first package
To install your first package, use `flox install`:

```
% flox install boost
```

> The first time you install a package, Flox will ask you whether you want to create a default environment. If you choose to do so, Flox will then ask you whether you want to configure your dotfiles to automatically activate the new default environment.

Reviewer: ```suggestion
The first time you install a package, Flox will ask you whether you want to create a default environment. If you choose to do so, Flox will then ask you whether you want to configure your shell to automatically activate the new default environment.
```

Dotfiles is a general term for config for all kinds of tools, and we should make it clear that we're only touching the config for this shell. People get antsy about tools touching their configs, so I like to make it clear that we're doing the absolute minimum.

```
[show the flow]
```

> If you do not choose for this to be automated at the time of your first package installation, you can [follow these instructions]() to add Flox to your dotfiles manually.

Reviewer: ```suggestion
If you do not choose for this to be automated at the time of your first package installation, you can [follow these instructions]() to add Flox to your shell config manually.
```

### Verifying configuration

> Exit your active shell and create a new one, causing the dotfile changes to take effect. The first time this happens, you may experience a short delay while your default environment is materialized.

Reviewer: ```suggestion
Exit your active shell and create a new one, causing the dotfile changes to take effect. The first time this happens, you may experience a short delay while your default environment is materialized. The next time you open a shell it should be quick because the environment has been cached.
```

> Once the shell is available, verify that your default environment is active by running `flox` with no arguments:

Reviewer: A shorter way to do this would be to either run `flox list` to see the packages available in the default environment, or run `flox envs list`.

> 

Reviewer 2: This is totally fine to use, but might be better to suggest using `flox envs` because this will show them if they have the default env, even if it's not active.  And then they learn that command.

```
% flox
flox version 1.3.6

Usage: flox OPTIONS (init|activate|search|install|...) [--help]

Use 'flox --help' for full list of commands and more information

✨ Active environments:
  default           /Users/floxfan
```

If you see `default` listed amongst the active environments, your dotfiles have been correctly modified. The default environment will be active whenever you log in.

Verify that your package has been installed using `flox list`:

```
% flox list
boost: boost (boost@1.81.0)
```

> ## Create project environments

Reviewer: I think this section can be shorter since this is a migration guide and not strictly about project environments.

As you continue to migrate packages from Homebrew to Flox, you may find that you don’t need them all in your default environment.

The default environment is intended for packages that should be available to the user across all of the contexts where they work. It is commonly used for general utilities like `gh`, `gnused`, and `curl` that apply to many situations.

> Packages that are required for specific contexts (e.g., different projects in a monorepo, different repos, different customer deployments, etc.) are often installed into project environments which are then activated in addition to the default environment.

Reviewer 2: We don't talk about "project environments" as a concept anywhere, so maybe consider dropping that as an adjective. Instead, could say just "environments" and in the section title "Create environments for projects".

The concept page for environments uses the terms "path" and "centrally managed" to talk about the use cases for environments, and I think in this case you're referencing the "path" use case. My goal is to reduce the number of terms a user has to understand.

> A project environment can be created using `flox init`:

Reviewer 2: Maybe link to the creating environments tutorial and remove the CLI output here?

> A project environment can be created using `flox init`:

Reviewer: Regarding the previous comment, I think everything from here to the next section might want to be replaced with links to other documentation pages.
